### PR TITLE
kvserver: ProcessSplitAfterRightHandSideHasBeenRemoved with all lease types

### PR DESCRIPTION
**Pull request by iskettaneh**
This commit allows the test TestProcessSplitAfterRightHandSideHasBeenRemoved to run with all lease types.

References: #133763

Release note: None

### CODE REVIEW

**Review by arulajmani — COMMENTED** [review]
Mostly looks good. While you're here, do you also want to address the other tests that use `setupPartitionedRange`? I think there's just 4 -- I'd say let's do it in a separate PR, once this lands, but maybe we can batch multiple commits in that PR if it's easier.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-:-OCeeN4F-cXQtVr39FXe:b-gg3u1x)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/kv/kvserver/client_raft_helpers_test.go` line 389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCecgEx0dHDa_QSpTf5:-OCecgEx0dHDa_QSpTf6:b-qdsauu) ([raw file](https://github.com/cockroachdb/cockroach/blob/373ac59e0fd4fd469efe0ab6b4786c6740a37c10/pkg/kv/kvserver/client_raft_helpers_test.go#L389)):*
> ```Go
> 		// Also partition the store in the storeliveness layer.
> 		targetStore, err := tc.Servers[partitionedNodeIdx].GetStores().(*kvserver.Stores).
> 			GetStore(tc.Servers[partitionedNodeIdx].GetFirstStoreID())
> ```

Can you use `GetFirstStoreFromServer` here instead?
___
*[`pkg/kv/kvserver/client_raft_helpers_test.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCecrOt0QFT568t-Klz:-OCecrOuA9BoIhP99MP4:bup2id4) ([raw file](https://github.com/cockroachdb/cockroach/blob/373ac59e0fd4fd469efe0ab6b4786c6740a37c10/pkg/kv/kvserver/client_raft_helpers_test.go#L402)):*
> ```Go
> 		}
> 
> 		store, err := tc.Servers[s].GetStores().(*kvserver.Stores).
> ```

Same as above -- can we use `GetFirstStoreFromServer` here as well? 
___
*[`pkg/kv/kvserver/client_raft_helpers_test.go` line 415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCed2V19nkISFVfAW59:-OCed2V19nkISFVfAW5A:b-76yc3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/373ac59e0fd4fd469efe0ab6b4786c6740a37c10/pkg/kv/kvserver/client_raft_helpers_test.go#L415)):*
> ```Go
> 						drop := shouldDropStoreLivenessMessage(msg.From.StoreID, msg.To.StoreID)
> 						if drop {
> 							log.Infof(context.Background(), "dropping msg %s from store %d: to %d",
> ```

nit: "dropping StoreLiveness msg %s"

Otherwise, this will read as "dropping msg MsgHeartbeat" which is confusing -- spent ~10 minutes yesterday wondering why raft heartbeats were enabled based on this log line yesterday, before realizing it was because store liveness relted.
___
*[`pkg/kv/kvserver/client_raft_helpers_test.go` line 418 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCedLvNE2UrHJoVFCU0:-OCedLvNE2UrHJoVFCU1:b-fhlpo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/373ac59e0fd4fd469efe0ab6b4786c6740a37c10/pkg/kv/kvserver/client_raft_helpers_test.go#L418)):*
> ```Go
> 								msg.Type, msg.From.StoreID, msg.To.StoreID)
> 						} else {
> 							log.Infof(context.Background(), "allowing msg %s from store %d: to %d",
> ```

nit: same as above -- let's reword this to say "allowing StoreLiveness ..."
___
*[`pkg/kv/kvserver/client_raft_test.go` line 5328 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCedY6i75xVD2KunG8l:-OCedY6i75xVD2KunG8m:bfff1z6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5328)):*
> ```Go
> 				st := cluster.MakeTestingClusterSettings()
> 				kvserver.OverrideDefaultLeaseType(ctx, &st.SV, leaseType)
> 				kvserver.ExpirationLeasesOnly.Override(ctx, &st.SV, false) // override metamorphism
> ```

Should we get rid of this override for expiration based leases? This should now be handled by `OverrideDefaultLeaseType`. Infact, as-is, we'll be using epoch based leases for the lease-type=LeaseExpiration variant.
___
*[`pkg/kv/kvserver/client_raft_test.go` line 5333 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCedq8h7dpoJdaJ6ylm:-OCedq8h7dpoJdaJ6yln:bbylmt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5333)):*
> ```Go
> 					// can remove this. Leader leases require us to reject lease requests on
> 					// replicas that are not the leader.
> 					kvserver.RejectLeaseOnLeaderUnknown.Override(ctx, &st.SV, true)
> ```

Could you confirm that we do indeed need this?
___
*[`pkg/kv/kvserver/client_raft_test.go` line 5543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCeeGBA1VKQdc-0G4pE:-OCeeGBA1VKQdc-0G4pF:bghxr81) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5543)):*
> ```Go
> 
> 			if leaseType == roachpb.LeaseLeader {
> 				// Since both LHS and RHS use the same store, let's remove the store
> ```

Why do we need this?

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136237)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/client_raft_helpers_test.go` line 389 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCecgEx0dHDa_QSpTf5:-OCejBWi91HCIGXk5DjI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/373ac59e0fd4fd469efe0ab6b4786c6740a37c10/pkg/kv/kvserver/client_raft_helpers_test.go#L389)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can you use `GetFirstStoreFromServer` here instead?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_raft_helpers_test.go` line 402 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCecrOt0QFT568t-Klz:-OCejHg390dremWOn1NS:b-k7r0u6) ([raw file](https://github.com/cockroachdb/cockroach/blob/373ac59e0fd4fd469efe0ab6b4786c6740a37c10/pkg/kv/kvserver/client_raft_helpers_test.go#L402)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Same as above -- can we use `GetFirstStoreFromServer` here as well?
</blockquote></details>

We are actually using the store below, not just the id
___
*[`pkg/kv/kvserver/client_raft_test.go` line 5328 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCedY6i75xVD2KunG8l:-OCemG9d8vtAB880KcSn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5328)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we get rid of this override for expiration based leases? This should now be handled by `OverrideDefaultLeaseType`. Infact, as-is, we'll be using epoch based leases for the lease-type=LeaseExpiration variant.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_raft_test.go` line 5333 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCedq8h7dpoJdaJ6ylm:-OCejyQY8B92KH0-qVUR:b-pogsro) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5333)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could you confirm that we do indeed need this?
</blockquote></details>

The test passes both ways. This got copied over from another PR. We don't need it anymore?
___
*[`pkg/kv/kvserver/client_raft_test.go` line 5543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCeeGBA1VKQdc-0G4pE:-OCek6QZ8eHOqVmiJ6AX:b4x3m37) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5543)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Why do we need this?
</blockquote></details>

The test partitions two ranges, and we partition the store liveness layer twice! The problem is that at some point during the test, it resolves a partition for one range, and expects that range to be working fine (including store liveness stuff). Therefore, I had to centralize the store liveness partition to one object form 2/4 of the subtests inside this test.

**Review by arulajmani — COMMENTED** [review]
Should be there after this round. 

> Reviewed 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-:-OCi7jJ9DEaEoYyIklyA:b-5bcd59)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/kv/kvserver/client_raft_test.go` line 5333 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCedq8h7dpoJdaJ6ylm:-OCi6hUZ02LevWoSzmso:b-vgyi4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5333)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

The test passes both ways. This got copied over from another PR. We don't need it anymore?
</blockquote></details>

Yeah, let's remove it if it's not needed. 
___
*[`pkg/kv/kvserver/client_raft_test.go` line 5543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCeeGBA1VKQdc-0G4pE:-OCi7Ace0rMbM4KpGWCF:b-xvou8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5543)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

The test partitions two ranges, and we partition the store liveness layer twice! The problem is that at some point during the test, it resolves a partition for one range, and expects that range to be working fine (including store liveness stuff). Therefore, I had to centralize the store liveness partition to one object form 2/4 of the subtests inside this test.
</blockquote></details>

Ack. I wonder if it's easier to resolve the partition on both ranges instead of doing this -- that seems more natural. There should never be a case where there are two ranges, with the same replica set, where one range is partitioned and another isn't.

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136237)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/client_raft_test.go` line 5333 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCedq8h7dpoJdaJ6ylm:-OCivliXDKC-3IhmOme_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5333)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Yeah, let's remove it if it's not needed.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_raft_test.go` line 5543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCeeGBA1VKQdc-0G4pE:-OCivmSWBURw8vC74lLb:b-5u9skx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5543)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Ack. I wonder if it's easier to resolve the partition on both ranges instead of doing this -- that seems more natural. There should never be a case where there are two ranges, with the same replica set, where one range is partitioned and another isn't.
</blockquote></details>

Removed it from one testcase, but had to leave it in another testcase. The test depends on the fact that one range is partitioned while another range on the same node isn't partitioned. I left a TODO to fix that.

**Review by arulajmani — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-:-OCjRR3H1VKs0HIyix3U:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/kv/kvserver/client_raft_test.go` line 5543 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136237#-OCeeGBA1VKQdc-0G4pE:-OCjROH5EwDvN7A8qi9i:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e94a760bbfc681b8fce62d84fbd0e2412c32917c/pkg/kv/kvserver/client_raft_test.go#L5543)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Removed it from one testcase, but had to leave it in another testcase. The test depends on the fact that one range is partitioned while another range on the same node isn't partitioned. I left a TODO to fix that.
</blockquote></details>

SGTM

**Comment by iskettaneh** [comment]
TFTR!

bors r+
